### ISSUE TITLE
HMAC_Init_ex failed to clean up the key buffer, leaving the HMAC key in plaintext

### ISSUE THREAD

**Issue by lantse-cap1** [openssl/openssl#10743]
When HMAC_Init_ex() returned, the secret key for HMAC was kept in the `key[HMAC_MAX_MD_CBLOCK_SIZE]` inside HMAC_CTX (aka 'hmac_ctx_st') in plaintext, which is a security hole: hackers can obtain the secret key for HMAC by dumping the memory. In fact, this key buffer in HMAC_CTX was not used any more after the secretKey was supplied to HMAC_Init_ex(). 

In OpenSSL_1.0.2, we cleaned up the key buffer ourselves:
````
  HMAC_CTX_init(&m_ctx);
  HMAC_Init_ex( &m_ctx,  (const void*)(secretKey), keyLength,  HashAlgo(),   NULL);

  // erase key buffer inside hmac ctx
  memset(m_ctx.key, 0, m_ctx.key_length);
  m_ctx.key_length = 0;
````
However, because OpenSSL_1.1.1 makes the HMAC_CTX opaque, we cannot access the key buffer in HMAC_CTX like below,

````
  m_ctx = HMAC_CTX_new();
  HMAC_Init_ex(m_ctx,  (const void*)(secretKey),  keyLength, HashAlgo(), NULL);

  // erase key buffer inside hmac ctx
  memset(m_ctx->key, 0, m_ctx->key_length);
  m_ctx->key_length = 0;
````
we got  compilation `error: member access into incomplete type 'HMAC_CTX' (aka 'hmac_ctx_st')
  memset(m_ctx->key, 0, m_ctx->key_length);`

Our experiments show that to complete 100million HMAC digests using the same key,  init HMAC_CTX with secret key only once will be much faster.

Therefore, the key buffer in 'HMAC_CTX' (aka 'hmac_ctx_st') should be cleared up before HMAC_Init_ex() returns to keep good performance and avoid leaking the secret key.

**Comment by mattcaswell** [comment]
> Our experiments show that to complete 100million HMAC digests using the same key, init HMAC_CTX with secret key only once will be much faster.

Can you expand a bit more on this statement? How are you performing multiple HMAC operations using the same key if you init the HMAC_CTX only once?

**Comment by lantse-cap1** [comment]
@mattcaswell below is our source code using the same key to perform multiple HMAC operations:

```
int HmacDigest::init(volatile unsigned char* secretKey, unsigned int keyLength) {
  m_ctx_initialized = false;
  // Initialise context
  HMAC_CTX_init(&m_ctx);
  HMAC_Init_ex(
    &m_ctx,
    (const void*)(secretKey),
    keyLength,
    HashAlgo(),
    NULL);

  // erase key buffer inside hmac ctx
  memset(m_ctx.key, 0, m_ctx.key_length);
  m_ctx.key_length = 0;

  m_ctx_initialized = true;
  return SUCCESS;
}

int HmacDigest::Digest(
  const unsigned char* text,
  const unsigned int textLength,
  unsigned char* outputBuffer,
  const unsigned int outputBufferSize) {

  if (!m_ctx_initialized) return ERROR;

  unsigned int hmac_result_len = 0;
  if (HMAC_Update(&m_ctx, text, textLength)) {
    if (!HMAC_Final(&m_ctx, outputBuffer, &hmac_result_len)) {
      TRACE(ERROR, "KeyId %s Digest fail: HMAC_Final not successful", PrintKeyId());
      hmac_result_len = 0;
    }
  } else {
    TRACE(ERROR, "KeyId %s Digest fail: HMAC_Update not successful", PrintKeyId());
  }

  HMAC_Init_ex(&m_ctx, NULL, 0, NULL, NULL);
  return hmac_result_len;
}
```
The test code:
```
HmacDigest hmacObject;
hmacObject.init(secretKey, keyLength)

unsigned char inputText[inputBufferSize];
unsigned char *signature = (unsigned char *) malloc(resultBufferSize);
for (long i = 0; i < 100000000; i++)
{
   int inputLength = getInputString(inputText);
  hmacObject.digest (inputText, inputLength, signature, resultBufferSize);
  
}
```

In this way, 100million hmac completed in 50seconds. When we tested calling HMAC_Init_ex with the secretKey every time, it took 104seconds on the same computer for the same data. We also did consistency tests to compare results of these two ways, they are 100% match.

I did review the source code for HMAC, the key[] in the HMAC_CTX was indeed not used (and no need to be used) after HMAC_Init_ex() with secretKey is called. Therefore, the key[] in the HMAC_CTX  should be cleared before HMAC_Init_ex() returns, not leaving it there in plaintext for hackers to view.

If you want, I can provide a PR to fix this issue. Thanks.

**Comment by mattcaswell** [comment]
> If you want, I can provide a PR to fix this issue. Thanks.

I actually already have a fix locally. Will raise a PR tomorrow.

**Comment by mattcaswell** [comment]
A couple of notes about your description of the issue above:

````
  // erase key buffer inside hmac ctx
  memset(m_ctx.key, 0, m_ctx.key_length);
````
This method of cleansing the key is potentially ineffective depending on how clever the compiler is. Some compilers will optimise such memsets away, so it's possible that your 1.0.2 solution isn't doing what you think it is doing.

Additionally, clearing the key may have limited benefits in reality. Although the key itself is cleared the ctx still contains a digested form of the key for the lifetime of the ctx - and this digested form of the key is equally useful to an attacker as the original form (i.e. it could be used to create HMACs as if the attacker had the original key).

Nonetheless, it does seem sensible to avoid storing the original key where we don't have to and therefore I have a fix in #10747.

**Comment by lantse-cap1** [comment]
Noted and thanks!

Do you have an estimation when the PR will be merged?

**Comment by mattcaswell** [comment]
> Do you have an estimation when the PR will be merged?

Assuming no major objections from reviewers, I'd hope it will be merged in the next week or so.

**Comment by mattcaswell** [comment]
This is fixed in master but the fix needs an adjustment for the 1.1.1 branch. New PR coming up.

**Comment by mattcaswell** [comment]
Now fixed in 1.1.1 too. Closing.

**Comment by lantse-cap1** [comment]
@mattcaswell when is the next release for 1.1.1?

**Comment by mattcaswell** [comment]
There is no currently scheduled 1.1.1 release. These get done on an "as needed" basis. Typically every 3-4 months unless a security issue means one gets done earlier. The last release was nearly 4 months ago which might suggest another one soonish.

**Comment by mspncp** [comment]
> @mattcaswell when is the next release for 1.1.1?

Note: If you need this fix quickly for development, you can either checkout the current `OpenSSL_1_1_1-stable` branch using git, or download our latest [daily snapshot](https://www.openssl.org/source/snapshot).
